test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperAttributeListTest.cs

```diff
@@ -0,0 +1,661 @@
+                    { new[] { A, B, A2, C }, "aname", "av", new[] { new TagHelperAttribute("aname", "av"), B, C } },
+                    { new[] { B, A2, A }, "aname", set, new[] { B, set } },
+                    { new[] { B, A2, A }, "aname", "av", new[] { B, new TagHelperAttribute("aname", "av") } },
+                    { new[] { B, A2, A, C }, "AName", set, new[] { B, set, C } },
+                    { new[] { B, A2, A, C }, "AName", "av", new[] { B, new TagHelperAttribute("AName", "av"), C } },
+                    { new[] { A, A3 }, "AName", set, new[] { set } },
+                    { new[] { A, A3 }, "AName", "av", new[] { new TagHelperAttribute("AName", "av") } },
+                    { new[] { A3, A }, "aname", set, new[] { set } },
+                    { new[] { A3, A }, "aname", "av", new[] { new TagHelperAttribute("aname", "av") } },
+                    { new[] { A, A2, A3 }, "AName", set, new[] { set } },
+                    { new[] { A, A2, A3 }, "AName", "av", new[] { new TagHelperAttribute("AName", "av") } },
+                    { new[] { C, B, A3, A }, "AName", set, new[] { C, B, set } },
+                    { new[] { C, B, A3, A }, "AName", "av", new[] { C, B, new TagHelperAttribute("AName", "av") } },
+                    { new[] { A, A2, A3 }, "Set", set, new[] { A, A2, A3, set } },
+                    { new[] { A, A2, A3 }, "Set", "Set Value", new[] { A, A2, A3, set } },
+                    { new[] { A, A2, A3, B, C }, "Set", set, new[] { A, A2, A3, B, C, set } },
+                    { new[] { A, A2, A3, B, C }, "Set", "Set Value", new[] { A, A2, A3, B, C, set } },
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(StringIndexerSetData))]
+        public void StringIndexer_SetsAttributeAtExpectedLocation(
+            IEnumerable<TagHelperAttribute> initialAttributes,
+            string keyToSet,
+            TagHelperAttribute setValue,
+            IEnumerable<TagHelperAttribute> expectedAttributes)
+        {
+            // Arrange
+            var attributes = new TagHelperAttributeList(initialAttributes);
+
+            // Act
+            attributes[keyToSet] = setValue;
+
+            // Assert
+            Assert.Equal(expectedAttributes, attributes, CaseSensitiveTagHelperAttributeComparer.Default);
+        }
+
+        [Fact]
+        public void StringIndexer_SetsAttributeThrowsIfInvalidIndex()
+        {
+            // Arrange
+            var attributes = new TagHelperAttributeList(new[]
+            {
+                new TagHelperAttribute("A", "AV"),
+                new TagHelperAttribute("B", "BV")
+            });
+
+            // Act & Assert
+            var exception = Assert.Throws<ArgumentOutOfRangeException>("index", () =>
+            {
+                attributes[2] = new TagHelperAttribute("C", "CV");
+            });
+        }
+
+        [Fact]
+        public void StringIndexer_SetsAttributeThrowsIfUnmatchingKey()
```

Comment:
- make this a `[Theory]` or add more `[Fact]`s to cover inconsistent names when the list already contains a match or two.
- rename for above reason as well as "unmatching key" saying nothing about what the key should match.